src/security/ZeroSensitiveMemory.h

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_SECURITY_ZEROSENSITIVEMEMORY_H
+#define SQUID_SRC_SECURITY_ZEROSENSITIVEMEMORY_H
+
+#include <cstring>
+
+namespace Security {
+
+inline void
+ZeroSensitiveMemory(void *dst, const size_t len)
+{
+    if (!len)
+        return;
+
+    if (!dst)
+        return;
```

Comment:
We must not ignore requests to clear nil buffers with positive sizes, especially in a function meant to deal with "security sensitive" data. Such requests are often an indication of a serious bug in the calling code. We must at least throw or even assert. The old throwing code here was on the right path, but you removed it (not sure why).

Preservation of the old throwing code was one of the reasons I suggested adding a specific "tolerate zero-length nil dst" comment to the new `if (!len)` check. You have added that check (thank you), but not the comment (not sure why). Without that comment and without throwing, there would be no reason to have two `if` statements.

This suggestion should restore the earlier throwing behavior (while keeping the length check):

```suggestion
    if (!dst)
        throw TextException("attempt to clear a non-empty buffer via a nil pointer", Here());
```

Reply:
Sorry I misread your previous change request.

Reply:
Just updating this old change request based on the failed attempt to avoid inlining:

```suggestion
    assert(dst);
```